### PR TITLE
Change JNI to make CallChannel special

### PR DESCRIPTION
**Pull request by swankjesse**
Previously the JNI bindings supported arbitrary Java interfaces. With this
change only CallChannel is supported.

### CODE REVIEW

File: zipline/native/InboundCallChannel.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  JSValue thisPointer = JS_GetPropertyStr(context->jsContext, global, name.c_str());
+  auto property = JS_NewAtom(context->jsContext, "serviceNamesArray");
+
+  JSValue jsResult = JS_Invoke(context->jsContext, thisPointer, property, 0, NULL);
+  jobject javaResult = context->toJavaObject(env, jsResult, true);
+
+  JS_FreeAtom(context->jsContext, property);
+  JS_FreeValue(context->jsContext, jsResult);
+  JS_FreeValue(context->jsContext, thisPointer);
+  JS_FreeValue(context->jsContext, global);
```
Comment (JakeWharton) [inline zipline/native/InboundCallChannel.cpp]: Is calling `context->jsContext` 20 times free? Like is it just accessing a value in a struct on the stack or something? I just want to make sure it's not a pointer chase...

File: zipline/native/OutboundCallChannel.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+#include "OutboundCallChannel.h"
+#include "Context.h"
+#include "ExceptionThrowers.h"
+
+OutboundCallChannel::OutboundCallChannel(Context* c, JNIEnv* env, const char* name, jobject object,
+                                 JSValueConst proxy)
+    : context(c),
+      name(name),
+      javaThis(env->NewGlobalRef(object)),
+      callChannelClass(static_cast<jclass>(env->NewGlobalRef(env->FindClass("app/cash/zipline/internal/bridge/CallChannel")))),
```
Comment (JakeWharton) [inline zipline/native/OutboundCallChannel.cpp:26]: Add this type to ProGuard rules with a full `-keep`

File: zipline/native/InboundCallChannel.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  return static_cast<jobjectArray>(javaResult);
+}
+
+jobjectArray InboundCallChannel::invoke(Context *context, JNIEnv* env, jstring instanceName,
+                                   jstring funName, jobjectArray encodedArguments) const {
+  JSValue global = JS_GetGlobalObject(context->jsContext);
+  JSValue thisPointer = JS_GetPropertyStr(context->jsContext, global, name.c_str());
+  auto property = JS_NewAtom(context->jsContext, "invoke");
```
Comment (JakeWharton) [inline zipline/native/InboundCallChannel.cpp]: Are these functions all `@JsName`'d to ensure they're stable?

File: zipline/native/OutboundCallChannel.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+JSValue
+OutboundCallChannel::disconnect(JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv) {
+  auto context = reinterpret_cast<const Context*>(JS_GetRuntimeOpaque(JS_GetRuntime(ctx)));
+  if (!context) {
+    return JS_ThrowReferenceError(ctx, "Null Java Proxy");
+  }
+  auto channel = reinterpret_cast<const OutboundCallChannel*>(JS_GetOpaque(this_val, context->jsClassId));
+  if (!channel) {
+    return JS_ThrowReferenceError(ctx, "Null Java Proxy");
```
Comment (JakeWharton) [inline zipline/native/OutboundCallChannel.cpp]: Is this copy/pasta? I don't expect this to ever happen, but the message seems like maybe it was from the old code.

**Review by JakeWharton — APPROVED** [review]

File: zipline/native/InboundCallChannel.cpp (hunk 1 above)
Comment (swankjesse) [reply zipline/native/InboundCallChannel.cpp]: Fixed. I extracted a local. I suspect the difference is negligible but it’s there enough that it’s more readable the way.

File: zipline/native/OutboundCallChannel.cpp (hunk 2 above)
Comment (swankjesse) [reply zipline/native/OutboundCallChannel.cpp:26]: We don’t have a rules file yet so I’m punting to follow-up.
https://github.com/cashapp/zipline/issues/322

File: zipline/native/InboundCallChannel.cpp (hunk 3 above)
Comment (swankjesse) [reply zipline/native/InboundCallChannel.cpp]: Yes!

File: zipline/native/OutboundCallChannel.cpp (hunk 4 above)
Comment (swankjesse) [reply zipline/native/OutboundCallChannel.cpp]: Yeah, I didn’t attempt to make behavior-changes in this PR, and that means the word ‘proxy’ still exists too much. Fixed.

File: zipline/native/OutboundCallChannel.cpp (hunk 2 above)
Comment (JakeWharton) [reply zipline/native/OutboundCallChannel.cpp:26]: It's in `jniMain/resources/META-INF/proguard/`

**Pull request merged**